Magic/src/Magic/Abilities.hs

```diff
@@ -96,6 +96,14 @@ availableFromBattlefield rSelf you =
     _ -> false
 
 
+-- | Checks whether an activation's source is on the graveyard
+availableFromGraveyard :: Contextual (View Bool)
+availableFromGraveyard rSelf you =
+  case rSelf of
+    (Some (Graveyard you'), _) | you' == you ->
+      (== you) <$> view (asks (controller . objectBase rSelf))
```

Comment:
If you know that a card is in your graveyard, you don't need to do the controller check anymore: only cards that you own/control can be in your own graveyard.